### ISSUE TITLE
[rllib] sampling in asynchronous/real-time environments

### ISSUE THREAD

**Issue by opherlieber** [ray-project/ray#4625]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **Ray installed from (source or binary)**: source
- **Ray version**: master branch
- **Python version**: 3.6
- **Exact command to reproduce**:

### Describe the problem

Hi,

I'm using rllib on games without emulated environments, running in real-time. The frames arrive at a fixed 30 FPS, and sampling is typically 7.5 or 15 steps per second per worker (Using frame-skipping of 4 or 2). When there is a game-over it can take 1 or more seconds for a restart to happen which will block the env.reset() for that worker while the other workers continue sampling.
Any hiccups/delays between environment steps (not including env.reset()) will cause unintended/late actions and in many cases an unintended game-over and inability to learn well.

So far I have it working using one of the asynchronous optimizers (APEX/APPO) and the following config options:
- sample_async enabled
- worker_side_prioritization disabled for APEX (Otherwise I get a big hiccup every n_step at the worker)

This gives me a max step->step time of around ~60ms per ENV/worker which is good enough, and something I cannot achieve so easily, if at all, in any other framework.

I've noticed recently the 'sample_async' option has been marked as deprecated, so I am wondering how to solve my use-case. The sampling queue depth of '2' (Or optionally higher) in the async optimizers should achieve a similar effect as the sample_async option, but it's not currently the case due to 2 issues:

- The collect_metrics() call after each iteration is blocking across all workers. This cancels the sampling pipeline achieved by the sampling queue depth. Also, if one of the workers reaches a game-over (which is common especially at start of training) the get_metrics() call will stall until that ENV can restart, causing the other workers to stall as well for over a second and reach unintended game overs (Changing batch_mode to 'complete_episodes' may solve it, but this will cause extremely large idle times without any training between iterations as episodes become longer)
- Updating the weights in non-async sampling will stall the sampling during the weight updates. I found this takes ~60ms so it's not as critical as the previous issue but can still cause delays 

Is the deprecation of the sample_async final, and if yes is there any way to mitigate the get_metrics issue? For example to make it asynchronous somehow at least in the async optimizers (Maybe to continue calling optimizer.step() while waiting for all the metric results)

Thanks!

**Comment by ericl** [comment]
Thanks for bringing this use case up. I think that, given we cannot easily achieve the required latency in other settings, we will have to continue to support sample_async.

Feel free to open a PR to un-deprecate it. I think the config doc could also be updated to point out the main use case of sample_async -- the reason it was originally deprecated was because many users were confused about what it does.

**Comment by ericl** [comment]
Auto-closing stale issue.
